pandas/core/arrays/base.py

```diff
@@ -353,6 +371,66 @@ def unique(self):
         uniques = unique(self.astype(object))
         return self._constructor_from_sequence(uniques)
 
+    def _values_for_factorize(self):
+        """Return an array and missing value suitable for factorization.
+
+        Returns
+        -------
+        values : ndarray
+            An array suitable for factoraization. This should maintain order
+            and be a supported dtype. By default, the extension array is cast
+            to object dtype.
+        """
+        return self.astype(object)
+
+    def factorize(self, na_sentinel=-1):
+        """Encode the extension array as an enumerated type.
+
+        Parameters
+        ----------
+        na_sentinel : int, default -1
+            Value to use in the `labels` array to indicate missing values.
+
+        Returns
+        -------
+        labels : ndarray
+            An interger NumPy array that's an indexer into the original
+            ExtensionArray.
+        uniques : ExtensionArray
+            An ExtensionArray containing the unique values of `self`.
+
+            .. note::
+
+               uniques should *not* contain a value for the NA sentinel,
+               if values in `self` are missing.
+
+        See Also
+        --------
+        pandas.factorize : top-level factorize method that dispatches here.
+
+        Notes
+        -----
+        :meth:`pandas.factorize` offers a `sort` keyword as well.
+        """
+        from pandas.core.algorithms import _factorize_array
+        import pandas.core.dtypes.common as com
+        from pandas._libs.tslib import iNaT
+
+        mask = self.isna()
+        arr = self._values_for_factorize()
+
+        # Mask values going into the hash table with the appropriate
+        # NA type.
+        if com.is_signed_integer_dtype(arr):
+            arr[mask] = iNaT
+        elif com.is_float_dtype(arr) or com.is_object_dtype(arr):
+            arr[mask] = np.nan
+
+        labels, uniques = _factorize_array(arr, check_nulls=True,
```

Comment:
we could also use `pd.factorize` here if we want to limit the use of internal APIs

Reply:
I was just trying that. Unfortunately, I have to pass `check_nulls=True` for categorical to work properly. It's typically False for integers. But perhaps that's a sign we're doing something improper :/

Reply:
Ah, yes, because there we still know what the original dtype was ... 
Hmm, that maybe shows that we rather need a `na_value` keyword (just from a perspective of making `pd.factorize` generally usable, not necessarily in this PR). 
In general, the default of `na_value=None` could be the `val != val` idiom, and you can override that with `val == na_value` with this keyword (e.g. to specify iNaT or -1 for integer data).

Reply:
OK, this issue and the fact that UInt64Hashtable doesn't have a null value are making me come back around to the mask approach.

That seems like the only way we can ensure that we'll actually get things correct.